File: Backend/app/schemas/roadmap.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


class RoadmapStepBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    is_completed: bool = False
    order_index: int = Field(default=0, ge=0)


class RoadmapStepCreate(RoadmapStepBase):
    pass


class RoadmapStepUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    is_completed: Optional[bool] = None
    order_index: Optional[int] = Field(None, ge=0)


class RoadmapStep(RoadmapStepBase):
    id: int
    roadmap_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class RoadmapBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None


class RoadmapCreate(RoadmapBase):
    steps: Optional[List[RoadmapStepCreate]] = []


class RoadmapUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None


class Roadmap(RoadmapBase):
    id: int
    goal_id: int
    progress_percentage: float
    steps: List[RoadmapStep] = []
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: Frontend/src/types/roadmap.ts
export interface RoadmapStep {
  id: number;
  roadmap_id: number;
  title: string;
  description?: string;
  is_completed: boolean;
  order_index: number;
  created_at: string;
  updated_at: string;
}

export interface Roadmap {
  id: number;
  goal_id: number;
  title: string;
  description?: string;
  progress_percentage: number;
  steps: RoadmapStep[];
  created_at: string;
  updated_at: string;
}

export interface RoadmapCreate {
  title: string;
  description?: string;
  steps?: Omit<RoadmapStep, 'id' | 'roadmap_id' | 'created_at' | 'updated_at'>[];
}

export interface RoadmapUpdate {
  title?: string;
  description?: string;
}
